### PR TITLE
Add Prometheus /metrics-Endpoint

### LINKED ISSUE #692: web_server: Expose Prometheus /metrics-Endpoint
**Describe the problem you have/What new integration you would like**
Expose available data on /metrics in web_server-component to allow prometheus scraping data from the esphome node.

**Please describe your use case for this integration and alternatives you've tried:**
Prometheus (https://prometheus.io/) is one of the mostly used Monitoring and TSDB-Solutions around infrastructure. Data is collected by prometheus scraping an HTTP-Page - mostly under /metrics - periodically.
If esphome exposes it's metrics directly, esphome-nodes could be easily integrated into existing Prometheus-Environments, using both the TSDB and Alerting-Features. The need for an additional exporter (like going via MQTT or Home Assistant) would be dropped, reducing complexity.

**Additional context**
An example format could be:
```
light_brightness{id="neopixel_light",platform="neopixelbus"} 255
light_color{id="neopixel_light",channel="r",platform="neopixelbus"} 255
```
and the following for an sensor:
```
sensor_value{id="bathroom_temperature",platform="xiaomi_lywsdcgq"} 22.1
```

### PR DESCRIPTION
**Pull request by margau**
## Description:
This PR adds optional support for prometheus in the web_server-component.
At the moment, only sensors values and their state (Number/NaN) are supported - other components to follow.

This PR is not ready yet, but I'd be interested in some early feedback.

**Related issue (if applicable):** fixes https://github.com/esphome/feature-requests/issues/692

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#556

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/web_server/__init__.py [hunk 1]
```diff
@@ -49,3 +50,5 @@ def to_code(config):
         cg.add_define('WEBSERVER_JS_INCLUDE')
         with open(config[CONF_JS_INCLUDE], "r") as myfile:
             cg.add(var.set_js_include(myfile.read()))
+    if config[CONF_PROMETHEUS]:
+        cg.add_define('WEBSERVER_PROMETHEUS')
```
Comment (glmnet) [inline esphome/components/web_server/__init__.py:54]: Ok I believe we can have the #ifdef thing here

File: esphome/components/web_server/web_server.h [hunk 2]
```diff
@@ -132,6 +139,10 @@ class WebServer : public Controller, public Component, public AsyncWebHandler {
 
   /// Dump the light state as a JSON string.
   std::string light_json(light::LightState *obj);
+  /// Return the type for prometheus
```
Comment (glmnet) [inline esphome/components/web_server/web_server.h:142]: shouldn't these have the #ifdef too?

File: esphome/components/web_server/web_server.cpp [hunk 3]
```diff
@@ -504,6 +559,84 @@ std::string WebServer::light_json(light::LightState *obj) {
     obj->dump_json(root);
   });
 }
+
```
Comment (glmnet) [inline esphome/components/web_server/web_server.cpp:562]: for new whole methods please add a new file like web_server_prometeus.cpp I believe it should just work

**Review by glmnet — COMMENTED** [review]
Try adding a prometeus: true in some of the testn.yaml so may be some compile issue are tracked already.

File: esphome/components/web_server/web_server.cpp (hunk 3 above)
Comment (margau) [reply esphome/components/web_server/web_server.cpp]: Done, have moved everything into it's own class.

File: esphome/components/web_server/web_server.h (hunk 2 above)
Comment (margau) [reply esphome/components/web_server/web_server.h]: Should be resolved to the web_server_prometheus-class, which is only included when WEBSERVER_PROMETHEUS is set.

File: esphome/components/web_server/__init__.py (hunk 1 above)
Comment (margau) [reply esphome/components/web_server/__init__.py:54]: Could you clarify this? Not sure what meant exactly.

**Comment by margau** [comment]
From my perspective, this Feature is ready @glmnet

**Comment by glmnet** [comment]
Great!

**Pull request merged**